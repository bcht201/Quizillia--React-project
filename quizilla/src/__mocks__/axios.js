'use-strict';
module.exports = {
    get: () => {
        return Promise.resolve({
            data: [
                {
                    results: [
                    {
                        category: "Mythology",
                        type: "multiple",
                        difficulty: "hard",
                        question: "Talos, the mythical giant bronze man, was the protector of which island?",
                        correct_answer: "Crete",
                        incorrect_answers: [
                            "Sardinia",
                            "Sicily",
                            "Cyprus"
                            ]
                        }
                    ]
                }
            ]
        })
    }
}