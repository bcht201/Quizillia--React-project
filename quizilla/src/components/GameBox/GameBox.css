.questionContainer{
    display: flex;
    align-items: center;
    flex-direction: column;
    padding: 0 60px;
}

.answerButton{
    min-width: 50px;
    margin-bottom: 15px;
    padding: 4px;
}

h3 {
    margin-top: 2%;
    margin-bottom: 30px;
    padding: 0;
}

.buttonContainer{
    display: flex;
    flex-flow: wrap;
    flex-direction: row;
}