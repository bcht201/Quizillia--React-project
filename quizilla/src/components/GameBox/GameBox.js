import React from 'react'

const GameBox = (props) => {
    let answers = [...props.gameInfo.incorrect_answers, props.gameInfo.correct_answer].sort();  
    
    const onClickHandler = (event) => {
        // console.log(event.target.value);
        props.calculateScore(event.target.value);
    }

    return(
        <div>
            <h3>{props.gameInfo.question}</h3>
            
            {answers.map(answer => {
                return <button value = {answer} id = {answer} onClick={(event)=> onClickHandler(event)}>{answer}</button>
            })}
        </div>
    )
}

export default GameBox