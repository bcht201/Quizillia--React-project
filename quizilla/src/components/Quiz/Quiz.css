.currentPlayer {
    text-align: center;
    margin-top: 5%;
    font-size: 30px;
    color: green;
}