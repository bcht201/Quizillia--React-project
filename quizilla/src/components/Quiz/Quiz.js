import React from 'react'
import axios from 'axios'
import GameBox from '../GameBox/GameBox';
import './Quiz.css'

class Quiz extends React.Component{
    constructor(props) {
        super(props);
        this.state= {
            data: [],
            index: 0,
            score: [],
            numberOfPlayers: Number(this.props.players),
            counter: 0
            
        }
    }

    componentDidMount = () => {
        this.createScoreArray()
        this.apiCall();
    }

    createScoreArray = () => {
        this.setState({score: Array(this.state.numberOfPlayers).fill(0)})
    }

    apiCall = () => {
        axios.get(`https://opentdb.com/api.php?amount=${this.props.maxNumberOfQuestions}&category=${this.props.catID}&difficulty=${this.props.diff}`)
        .then(response => {
            this.setState({data: response.data.results})})
    }

    // calculateScore = answer =>{
    //     let points;
    //     if(answer === this.state.data[this.state.index].correct_answer){
    //         points = 1;
    //     } else{
    //         points = -1;
    //     }
    //     this.setState(prevState => {
    //         let newScore = [...prevState.score];
    //         let questionIncrement;
    //         if(this.state.numberOfPlayers === 1) {
    //             questionIncrement = 1;
    //         } else {
    //             questionIncrement = prevState.counter % this.state.numberOfPlayers
    //         }
    //         newScore[this.state.counter % this.state.numberOfPlayers] += points;
    //         return ({score: newScore, counter: prevState.counter + 1, index: prevState.index + questionIncrement})})
    // }

    nextGameState = answer =>{
        let newScore = this.calculateScore(answer);
        let nextQuestion = this.updateQuestion();
        this.setState({score : newScore, counter: this.state.counter + 1, index: this.state.index + nextQuestion});
    }

    calculateScore = (answer) =>{
        let points = this.state.score;
        let currentPlayer = this.state.counter % this.state.numberOfPlayers;
        if(answer === this.state.data[this.state.index].correct_answer){
            points[currentPlayer] += 1;
        } else{
            points[currentPlayer] += -1;
        }
        return points;
    };

    updateQuestion = () =>{
        let questionIncrement;
        if(this.state.numberOfPlayers === 1) {
            questionIncrement = 1;
        } else {
            questionIncrement = this.state.counter % this.state.numberOfPlayers
        }
        return questionIncrement;
    }

    updateScore = () =>{

    }

    gameState = () => {
        if(this.state.index === this.props.maxNumberOfQuestions) {
            return this.generateResults()
        }
        else if(this.state.data.length !== 0) { 
            return this.generateNextQuestion()
        }
        else {
            return <p>LOADING</p>;
       }
    }

    generateResults = () => {
        let emojis = [<p>&#128563;</p>, <p>&#128513;</p> ];
        return this.state.score.map((score, index)=> {
            return(
                <p className = "resultsMsg"> Player {index+1} got {score}/{this.props.maxNumberOfQuestions}{score < 0? emojis[0] : emojis[1]} </p>
            )
        })
    }

    generateNextQuestion = () => {
        return( 
            <div>
                <p className = 'currentPlayer'>Current Player: {this.state.counter % this.state.numberOfPlayers + 1}</p>
                <GameBox gameInfo={this.state.data[this.state.index]} calculateScore = {this.nextGameState}/>
            </div>) 
    }

    render() {
        return (
            <div>
                {this.gameState()}
            </div>
            
        )
    }
}

export default Quiz;