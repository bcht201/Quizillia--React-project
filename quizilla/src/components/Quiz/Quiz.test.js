import React from 'react';
import {
    shallow, 
    mount
} from 'enzyme';
import Quiz from './Quiz'
import GameBox from '../GameBox/GameBox'
import axios from 'axios'
import renderer from 'react-test-renderer'

jest.mock('axios')

describe('Quiz.js', () => {
    let wrapper;

    beforeEach(() => {
      wrapper= shallow(<Quiz players={"1"}/>);
    });

    it('should update state for number of players on mount', ()=> {
        const instance = mount(<Quiz players={"1"}/>).instance();
        const spy = jest.spyOn(instance, 'createScoreArray');
        instance.componentDidMount();
        expect(spy).toHaveBeenCalled();
    });
    
    
    it('should have a div', () => {
        expect(wrapper.find('div').length).toEqual(1);
    });
    
    it('should not render GameBox', () => {
        expect(wrapper.find('GameBox').length).toEqual(0);
    });
    
    it('should match snapshot', () => {
        const tree = renderer.create(<Quiz players={"1"} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it('score state should have a length of 1', () => {
        expect(wrapper.state().score.length).toEqual(1);
    });

    it('testing GameStates if statement', () => {
        let wrapper2 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>)
        let fakeFxn = wrapper2.instance().generateResults = jest.fn();
        wrapper2.update();
        wrapper2.setState({index: 1 });
        expect(fakeFxn).toHaveBeenCalled();
    });

    it('testing GameStates else statement', () => {
        let wrapper3 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>)
        let fakeFxn = wrapper3.instance().generateNextQuestion = jest.fn();
        wrapper3.update();
        wrapper3.setState({data: [0, 0] });
        expect(fakeFxn).toHaveBeenCalled();
    });

    it('should return + 1 if the answer is correct', () =>{
        let wrapper3 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>);
        let mockData = [{
            "category": "Mythology",
            "type": "multiple",
            "difficulty": "hard",
            "question": "Talos, the mythical giant bronze man, was the protector of which island?",
            "correct_answer": "Crete",
            "incorrect_answers": [
                "Sardinia",
                "Sicily",
                "Cyprus"
            ]
        }];
        wrapper3.setState({score: [0], counter: 0, numberOfPlayers: 1, data: mockData, index: 0});
        wrapper3.update();
        let fakeCalcScore = wrapper3.instance().calculateScore ('Crete');
        expect(fakeCalcScore[0]).toEqual(1);
    });

    it('should return + 1 if the answer is correct', () =>{
        let wrapper3 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>);
        let mockData = [{
            "category": "Mythology",
            "type": "multiple",
            "difficulty": "hard",
            "question": "Talos, the mythical giant bronze man, was the protector of which island?",
            "correct_answer": "Crete",
            "incorrect_answers": [
                "Sardinia",
                "Sicily",
                "Cyprus"
            ]
        }];
        wrapper3.setState({score: [0], counter: 0, numberOfPlayers: 1, data: mockData, index: 0});
        wrapper3.update();
        let fakeCalcScore = wrapper3.instance().calculateScore ('Sicily');
        expect(fakeCalcScore[0]).toEqual(-1);
    });

    it('should return 1 if there is one player', () =>{
        let wrapper3 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>);
        let fakeUpdateQ = wrapper3.instance().updateQuestion();
        expect(fakeUpdateQ).toEqual(1);
    });

    it('should return 0 if there are 2 players and it is the first players turn', () =>{
        let wrapper3 = shallow(<Quiz players={'2'} maxNumberOfQuestions={1}/>);
        wrapper3.setState({counter: 0});
        wrapper3.update();
        let fakeUpdateQ = wrapper3.instance().updateQuestion();
        expect(fakeUpdateQ).toEqual(0);
    });
    
    it('should return 1 if there are 2 players and it is the second players turn', () =>{
        let wrapper3 = shallow(<Quiz players={'2'} maxNumberOfQuestions={1}/>);
        wrapper3.setState({counter: 1});
        wrapper3.update();
        let fakeUpdateQ = wrapper3.instance().updateQuestion();
        expect(fakeUpdateQ).toEqual(1);
    });

    it('gameState should run newScore, followed by nextQuestion', () =>{
        let wrapper3 = shallow(<Quiz players={'1'} maxNumberOfQuestions={1}/>);
        let mockData = [{
            "category": "Mythology",
            "type": "multiple",
            "difficulty": "hard",
            "question": "Talos, the mythical giant bronze man, was the protector of which island?",
            "correct_answer": "Crete",
            "incorrect_answers": [
                "Sardinia",
                "Sicily",
                "Cyprus"
            ]
        }];
        wrapper3.setState({score: [0], counter: 0, numberOfPlayers: 1, data: mockData, index: 0});
        wrapper3.update();
        let fakeCalcScore = wrapper3.instance().calculateScore = jest.fn();
        let fakeUpdateQ = wrapper3.instance().updateQuestion = jest.fn();
        wrapper3.instance().nextGameState();
        expect(fakeCalcScore).toHaveBeenCalled();
        expect(fakeUpdateQ).toHaveBeenCalled();
    })
}) 




