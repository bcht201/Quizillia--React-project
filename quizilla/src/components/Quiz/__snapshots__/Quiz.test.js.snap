// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz.js should match snapshot 1`] = `
<div>
  <p>
    LOADING
  </p>
</div>
`;
